system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -149,28 +534,43 @@ class CLibXpathContext : public CInterface, public IXpathContext
         return evaluateAsString(evaluate(clibCompiledXpath->getCompiledXPathExpression(), compiledXpath->getXpath()), evaluated, compiledXpath->getXpath());
     }
 
+    virtual double evaluateAsNumber(ICompiledXpath * compiledXpath) override
+    {
+        CLibCompiledXpath * clibCompiledXpath = static_cast<CLibCompiledXpath *>(compiledXpath);
+        if (!clibCompiledXpath)
+            throw MakeStringException(XPATHERR_MissingInput,"XpathProcessor:evaluateAsNumber: Error: Missing compiled XPATH");
+        return evaluateAsNumber(evaluate(clibCompiledXpath->getCompiledXPathExpression(), compiledXpath->getXpath()), compiledXpath->getXpath());
+    }
+
 private:
-    virtual bool setXmlDoc(const char * xmldoc) override
+    bool setContextDocument(xmlDocPtr doc, xmlNodePtr node)
     {
-        if (xmldoc && * xmldoc)
+        WriteLockBlock rblock(m_rwlock);
+
+        m_xmlDoc = doc;
+        m_xpathContext = xmlXPathNewContext(m_xmlDoc);
+        if(m_xpathContext == nullptr)
         {
-            m_xmlDoc = xmlParseDoc((const unsigned char *)xmldoc);
-            if (m_xmlDoc == nullptr)
-            {
-                ERRLOG("XpathProcessor:setxmldoc Error: Unable to parse XMLLib document");
-                return false;
-            }
+            ERRLOG("XpathProcessor:setxmldoc: Error: Unable to create new XMLLib XPath context");
+            return false;
+        }
 
-            // Create xpath evaluation context
-            m_xpathContext = xmlXPathNewContext(m_xmlDoc);
-            if(m_xpathContext == nullptr)
-            {
-                ERRLOG("XpathProcessor:setxmldoc: Error: Unable to create new XMLLib XPath context");
-                return false;
-            }
-            return true;
+        //relative paths need something to be relative to
+        xmlXPathSetContextNode(node, m_xpathContext);
+        xmlXPathRegisterVariableLookup(m_xpathContext, variableLookupFunc, this);
+        return true;
+    }
+    virtual bool setXmlDoc(const char * xmldoc) override
```

Comment:
the setXmlDoc is a public method in the interface, but is private in the class, is it intended?

Reply:
Hmmm don't know why that was done, but I agree.  I'll change it.